lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -417,135 +424,149 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
+  if (!disableMemRandomization || !disableRegRandomization) {
+    constexpr unsigned randomWidth = 32;
+    sv::RegOp randomMemReg;
+    b.create<sv::IfDefOp>("SYNTHESIS", std::function<void()>(), [&]() {
+      sv::RegOp randReg;
+      StringRef initvar;
+
+      // Declare variables for use by memory randomization logic.
+      if (!disableMemRandomization) {
+        b.create<sv::IfDefOp>("RANDOMIZE_MEM_INIT", [&]() {
+          initvar = moduleNamespace.newName("initvar");
+          b.create<sv::VerbatimOp>("integer " + Twine(initvar) + ";\n");
+          randomMemReg = b.create<sv::RegOp>(
+              b.getIntegerType(llvm::divideCeil(mem.dataWidth, randomWidth) *
+                               randomWidth),
+              b.getStringAttr("_RANDOM_MEM"),
+              b.getStringAttr(moduleNamespace.newName("_RANDOM_MEM")));
+        });
+      }
 
-    // Declare variables for use by register randomization logic.
-    SmallVector<sv::RegOp> randRegs;
-    b.create<sv::IfDefOp>("RANDOMIZE_REG_INIT", [&]() {
-      signed totalWidth = 0;
-      for (sv::RegOp &reg : registers)
-        totalWidth += reg.getElementType().getIntOrFloatBitWidth();
-      while (totalWidth > 0) {
-        auto name = b.getStringAttr(moduleNamespace.newName(Twine("_RANDOM")));
-        randRegs.push_back(
-            b.create<sv::RegOp>(b.getIntegerType(randomWidth), name, name));
-        totalWidth -= randomWidth;
+      // Declare variables for use by register randomization logic.
+      SmallVector<sv::RegOp> randRegs;
+      if (!disableRegRandomization) {
+        b.create<sv::IfDefOp>("RANDOMIZE_REG_INIT", [&]() {
+          signed totalWidth = 0;
+          for (sv::RegOp &reg : registers)
+            totalWidth += reg.getElementType().getIntOrFloatBitWidth();
+          while (totalWidth > 0) {
+            auto name =
+                b.getStringAttr(moduleNamespace.newName(Twine("_RANDOM")));
+            randRegs.push_back(
+                b.create<sv::RegOp>(b.getIntegerType(randomWidth), name, name));
+            totalWidth -= randomWidth;
+          }
+        });
       }
-    });
 
-    b.create<sv::InitialOp>([&]() {
-      b.create<sv::VerbatimOp>("`INIT_RANDOM_PROLOG_");
-
-      // Memory randomization logic.  The entire memory is randomized.
-      b.create<sv::IfDefProceduralOp>("RANDOMIZE_MEM_INIT", [&]() {
-        std::string verbatimForLoop;
-        llvm::raw_string_ostream s(verbatimForLoop);
-        s << "for (" << initvar << " = 0; " << initvar << " < " << mem.depth
-          << "; " << initvar << " = " << initvar << " + 1) begin\n"
-          << "  {{0}} = ";
-        auto repetitionCount = llvm::divideCeil(mem.dataWidth, randomWidth);
-        if (repetitionCount > 1)
-          s << "{";
-        for (size_t i = 0; i != repetitionCount; ++i) {
-          if (i > 0)
-            s << ", ";
-          s << "{`RANDOM}";
+      b.create<sv::InitialOp>([&]() {
+        b.create<sv::VerbatimOp>("`INIT_RANDOM_PROLOG_");
+
```

Comment:
Possibly we can remove initial block and verbatim `INIT_RANDOM_PROLOG`?

Reply:
We will omit them if both mem and reg randomization is disabled, e.g., with `--disable-all-randomization`.  I'd expect that's the main option that'll be used (of the new options).

Reply:
Make sense, I was missing early exit. Thanks!